Add invoketarget & invokeaction attributes

keithamus opened the pull request:
#### de91640478c442925d7cd283f91f1d1e334d1d77
<pre>
Add invoketarget &amp; invokeaction attributes

<a href="https://bugs.webkit.org/show_bug.cgi?id=262850">https://bugs.webkit.org/show_bug.cgi?id=262850</a>

Reviewed by Tim Nguyen.

This adds support for the experimental `invoketarget` and `invokeaction`
attributes, as specified in the open-ui &quot;Invokers&quot; explainer.

(<a href="https://open-ui.org/components/invokers.explainer/)">https://open-ui.org/components/invokers.explainer/)</a>

The `invoketarget` attribute maps to the IDL `invokeTargetElement`,
similar to `popoverTargetElement`, and the `invokeaction` is a freeform
string.

The Button behaviour checks for `invokeTargetElement` in its activation
behaviour, and dispatches an `InvokeEvent` if there is one.

This also adds some basic scaffolding for `handleInvokeInternal` which
will allow elements to provide their own invocation action algorithms.

Co-authored-by: Keith Cirkel &lt;keithamus@users.noreply.github.com&gt;

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/idlharness.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invokeelement-interface.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invokeevent-dispatch-shadow.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invokeevent-interface.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-button-event-dispatch.tentative-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isElementReflectionAttribute):
* Source/WebCore/dom/Event.h:
(WebCore::Event::isInvokeEvent const):
* Source/WebCore/dom/EventNames.h:
* Source/WebCore/dom/EventNames.in:
* Source/WebCore/dom/InvokeEvent.cpp: Added.
(WebCore::InvokeEvent::InvokeEvent):
(WebCore::InvokeEvent::create):
(WebCore::InvokeEvent::createForBindings):
(WebCore::InvokeEvent::eventInterface const):
(WebCore::InvokeEvent::isInvokeEvent const):
(WebCore::InvokeEvent::invoker const):
* Source/WebCore/dom/InvokeEvent.h: Added.
* Source/WebCore/dom/InvokeEvent.idl: Added.
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::defaultEventHandler):
* Source/WebCore/html/HTMLButtonElement.idl:
* Source/WebCore/html/HTMLElement.h:
(WebCore::HTMLElement::handleInvokeInternal):
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::invokeTargetElement const):
(WebCore::HTMLFormControlElement::invokeAction const):
(WebCore::HTMLFormControlElement::setInvokeAction):
(WebCore::HTMLFormControlElement::handleInvokeAction):
* Source/WebCore/html/HTMLFormControlElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::defaultEventHandler):
* Source/WebCore/html/HTMLInputElement.idl:
* Source/WebCore/html/InvokerElement.idl: Added.

Canonical link: <a href="https://commits.webkit.org/271428@main">https://commits.webkit.org/271428@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/45478424edaab638033b7f0589a7a141c6785478

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25922 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4532 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27202 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28021 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23737 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26239 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6294 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1960 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23819 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26173 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3419 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22330 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3039 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3038 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23284 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28601 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/22573 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23653 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23661 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29351 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/25142 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3066 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1277 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27226 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4456 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/32586 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6797 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3522 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7079 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3384 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Ahmad-S792 commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -3598,6 +3598,20 @@ InvisibleAutoplayNotPermitted:
 
+InvokeAttributeEnabled:
+  type: bool
+  status: stable
```
Shouldn't it be 'preview' for few releases?

keithamus replied:
🤦 yes thank you!

lukewarlow replied:
Or even just ~~"testing"~~ "testable" potentially?

Ahmad-S792 replied:
Yes - or even 'testable' till we have web-spec sorted out as @lukewarlow said.

Ahmad-S792 commented on metadata/contributors.json:
```diff
@@ -4502,6 +4502,16 @@
          "keishi"
       ]
    },
+   {
```
I think - if you can land this separately. It would be good. It will also give you option to use 'buildbot' (rebuild' button to restart failing bots (to check if there is any flaky test failures) etc.

nt1m commented on Source/WebKitLegacy/mac/WebView/WebPreferencesDefaultValues.mm:
```diff
@@ -207,6 +207,12 @@ bool defaultShouldConvertInvalidURLsToBlank()
     return shouldConvertInvalidURLsToBlank;
 }
 
+bool defaultInvokeAttributeEnabled()
```
You probably don't need this and `SDKAlignedBehavior::InvokeAttributeEnabled`. Popover needs this because the `popover` attribute caused compat issues (but not the target attributes), so this method disables it on older apps. I suspect invoke target attributes don't need that compat treatment though.

nt1m commented on Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
```diff
@@ -163,6 +163,11 @@ bool defaultCaptureAudioInUIProcessEnabled()
+bool defaultInvokeAttributeEnabled()
+{
+    return false;
+}
```
You can remove this from here and the header files, since the YAML file is no longer using it.

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1555,6 +1556,19 @@ PopoverState HTMLElement::popoverState() const
 }
 
+
+bool HTMLElement::FireInvokeEvent(Element* invoker, const AtomString& action)
```
```suggestion
bool HTMLElement::dispatchInvokeEvent(Element* invoker, const AtomString& action)
```
To match webkit style.

nt1m commented on Source/WebKitLegacy/mac/WebView/WebPreferencesDefaultValues.h:
```diff
@@ -65,6 +65,7 @@ bool defaultUseLegacyBackgroundSizeShorthandBehavior();
 bool defaultAllowDisplayOfInsecureContent();
 bool defaultAllowRunningOfInsecureContent();
 bool defaultShouldConvertInvalidURLsToBlank();
+bool defaultInvokeAttributeEnabled();
```
This needs to be removed too.

nt1m commented on Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:
```diff
@@ -59,6 +59,7 @@ enum class SDKAlignedBehavior {
     HTMLDocumentSupportedPropertyNames,
     InitializeWebKit2MainThreadAssertion,
     InspectableDefaultsToDisabled,
+    InvokeAttributeEnabled,
```
This is not needed.

nt1m commented on Source/WebCore/dom/InvokeEvent.cpp:
```diff
@@ -0,0 +1,75 @@
+{
+    auto* invoker = m_invoker.get();
+    if (LIKELY(is<Node>(currentTarget()))) {
+        auto& treeScope = downcast<Node>(currentTarget())->treeScope();
```
```suggestion
    if (RefPtr currentTarget = dynamicDowncast<Node>(currentTarget())) {
        auto& treeScope = currentTarget->treeScope();
```

nt1m commented on Source/WebCore/dom/InvokeEvent.cpp:
```diff
@@ -0,0 +1,75 @@
+
+RefPtr<Element> InvokeEvent::invoker() const
+{
+    auto* invoker = m_invoker.get();
```
No point caching `m_invoker` here. Just use it directly a few lines below.

nt1m commented on Source/WebCore/html/InvokerElement.idl:
```diff
@@ -0,0 +1,30 @@
+
+[EnabledBySetting=InvokeAttributeEnabled]
+interface mixin InvokerElement {
+    [CEReactions=NotNeeded, Reflect=invoketarget] attribute Element? invokeTargetElement;
+    [CEReactions=NotNeeded] attribute [AtomString] DOMString invokeAction;
```
bikeshedding for the spec, but I wonder if we should name these `invokesTargetElement` / `invokesAction` instead of `invoke`.

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2059,7 +2059,8 @@ static inline AtomString makeIdForStyleResolution(const AtomString& value, bool
     return (settings.ariaReflectionForElementReferencesEnabled() && name == HTMLNames::aria_activedescendantAttr)
-        || (settings.popoverAttributeEnabled() && name == HTMLNames::popovertargetAttr);
+        || (settings.popoverAttributeEnabled() && name == HTMLNames::popovertargetAttr)
+        || (settings.invokeAttributeEnabled() && (name == HTMLNames::invoketargetAttr || name == HTMLNames::invokeactionAttr));
```
`HTMLNames::invokeactionAttr` should be removed, it doesn't reflect to an `Element` property.

nt1m commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -3501,6 +3501,20 @@ InvisibleAutoplayNotPermitted:
+  status: testable
+  category: html
+  humanReadableName: "HTML invoke attribute"
+  humanReadableDescription: "Enable HTML invoke attribute support"
```
```suggestion
  humanReadableName: "HTML invoketarget & invokeaction attributes"
  humanReadableDescription: "Enable HTML invoketarget & invokeaction attribute support"
```

nt1m commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -3501,6 +3501,20 @@ InvisibleAutoplayNotPermitted:
     WebCore:
       default: false
 
+InvokeAttributeEnabled:
```
What do you think of this name?

```suggestion
InvokerAttributesEnabled:
```

nt1m reviewed (commented):
There's some iOS stuff to rebaseline, but in general this looks OK.

Can you change the commit title to "Add invoketarget & invokeaction attributes"

nt1m commented on Source/WebCore/html/HTMLElement.h:
```diff
@@ -159,6 +159,8 @@ class HTMLElement : public StyledElement {
     void setPopover(const AtomString& value) { setAttributeWithoutSynchronization(HTMLNames::popoverAttr, value); };
     void popoverAttributeChanged(const AtomString& value);
 
+    virtual void handleInvokeInternal(const AtomString& action) { (void)action; }
```
```suggestion
    virtual void handleInvokeInternal(const AtomString&) { }
```

nt1m commented on Source/WebCore/html/HTMLFormControlElement.cpp:
```diff
@@ -405,6 +406,61 @@ void HTMLFormControlElement::handlePopoverTargetAction() const
+    RefPtr element = dynamicDowncast<HTMLElement>(getElementAttribute(invoketargetAttr));
+    if (element)
+        return element.get();
+    return nullptr;
```
```suggestion
    return dynamicDowncast<HTMLElement>(getElementAttribute(invoketargetAttr));
```

nt1m commented on Source/WebCore/html/HTMLElement.h:
```diff
@@ -159,6 +159,8 @@ class HTMLElement : public StyledElement {
     void setPopover(const AtomString& value) { setAttributeWithoutSynchronization(HTMLNames::popoverAttr, value); };
     void popoverAttributeChanged(const AtomString& value);
 
+    virtual void handleInvokeInternal(const AtomString&) {}
```
```suggestion
    virtual void handleInvokeInternal(const AtomString&) { }
```
This needs a space to make the style bot happy.

nt1m reviewed (commented):
This looks generally good to me. @annevk WDYT of landing this?

lindseywild committed:
Add invoketarget & invokeaction attributes

https://bugs.webkit.org/show_bug.cgi?id=262850

Reviewed by Tim Nguyen.

This adds support for the experimental `invoketarget` and `invokeaction`
attributes, as specified in the open-ui "Invokers" explainer.

(https://open-ui.org/components/invokers.explainer/)

The `invoketarget` attribute maps to the IDL `invokeTargetElement`,
similar to `popoverTargetElement`, and the `invokeaction` is a freeform
string.

The Button behaviour checks for `invokeTargetElement` in its activation
behaviour, and dispatches an `InvokeEvent` if there is one.

This also adds some basic scaffolding for `handleInvokeInternal` which
will allow elements to provide their own invocation action algorithms.

Co-authored-by: Keith Cirkel <keithamus@users.noreply.github.com>

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/idlharness.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invokeelement-interface.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invokeevent-dispatch-shadow.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invokeevent-interface.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-button-event-dispatch.tentative-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isElementReflectionAttribute):
* Source/WebCore/dom/Event.h:
(WebCore::Event::isInvokeEvent const):
* Source/WebCore/dom/EventNames.h:
* Source/WebCore/dom/EventNames.in:
* Source/WebCore/dom/InvokeEvent.cpp: Added.
(WebCore::InvokeEvent::InvokeEvent):
(WebCore::InvokeEvent::create):
(WebCore::InvokeEvent::createForBindings):
(WebCore::InvokeEvent::eventInterface const):
(WebCore::InvokeEvent::isInvokeEvent const):
(WebCore::InvokeEvent::invoker const):
* Source/WebCore/dom/InvokeEvent.h: Added.
* Source/WebCore/dom/InvokeEvent.idl: Added.
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::defaultEventHandler):
* Source/WebCore/html/HTMLButtonElement.idl:
* Source/WebCore/html/HTMLElement.h:
(WebCore::HTMLElement::handleInvokeInternal):
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::invokeTargetElement const):
(WebCore::HTMLFormControlElement::invokeAction const):
(WebCore::HTMLFormControlElement::setInvokeAction):
(WebCore::HTMLFormControlElement::handleInvokeAction):
* Source/WebCore/html/HTMLFormControlElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::defaultEventHandler):
* Source/WebCore/html/HTMLInputElement.idl:
* Source/WebCore/html/InvokerElement.idl: Added.

Canonical link: https://commits.webkit.org/271428@main
